lib/Runtime/Executor/ThreadPoolExecutor.cpp

```diff
@@ -0,0 +1,377 @@
+  // done here instead of inside executeDAGNode() so that a node can be
+  // executed while placeholders are being propagated for the next node without
+  // the callback for that node deleting the execution state.
+  executionState->incrementInflightNodes((root->children).size());
+
+  // Convert the Context to shared_ptr.
+  std::shared_ptr<Context> sharedContext = std::move(context);
+
+  for (auto const &node : root->children) {
+    // Propagate placeholders from the given starter Context into the input
+    // Context for the current node being processed.
+    propagatePlaceholdersForNode(runId, node, sharedContext);
+
+    // Execute the node.
+    executeDAGNode(runId, node);
+  }
+}
+
+void ThreadPoolExecutor::propagatePlaceholdersForNode(
+    RunIdentifierTy runId, const DAGNode *node, std::shared_ptr<Context> ctx) {
+  // Get the execution state for the run.
+  auto it = executionStates_.find(runId);
+
+  if (it == executionStates_.end()) {
+    assert("Execution state should exist!");
+  }
+
+  std::shared_ptr<ExecutionState> executionState = it->second;
+
+  // Get the map of Placeholder -> Tensor mappings within ctx.
+  const Context::PlaceholderMap &phMap = ctx->pairs();
+
+  // Get the symbol table for the node.
+  const RuntimeBundle &bundle = node->runtimeBundle;
+  const SymbolTable &symbolTable = bundle.getSymbolTable();
+
+  for (const auto &symbolPair : symbolTable) {
+    const auto &symbolName = symbolPair.first;
+    const auto &symbolInfo = symbolPair.second;
+
+    if (symbolInfo.input) {
+      // If the symbol is an input, look for a Placeholder in ctx with
+      // the same name and copy the corresponding Tensor into the input Context
+      // being prepared for the node.
+      for (const auto &phTensorPair : phMap) {
+        auto *placeholder = phTensorPair.first;
+        const auto *tensor = phTensorPair.second;
+
+        if (symbolName == placeholder->getName()) {
+          executionState->insertIntoNodeCtx(node, placeholder, tensor->clone());
+        }
+      }
+    }
+  }
+}
+
+void ThreadPoolExecutor::executeDAGNode(RunIdentifierTy runId,
+                                        const DAGNode *node) {
+  // Get the execution state for the run.
+  auto executionStateIt = executionStates_.find(runId);
+
+  if (executionStateIt == executionStates_.end()) {
+    assert("Execution state should exist!");
+  }
+
+  std::shared_ptr<ExecutionState> executionState = executionStateIt->second;
+
+  // If execution has already failed due to another node, don't bother running
+  // this one.
+  if (executionState->getResultCode() == ResultCode::Failed) {
+    // Mark the node as no longer execution.
+    executionState->decrementInflightNodes();
+    return;
+  }
+
+  // Get the DeviceManager that can run the node.
+  auto deviceManagerIt = deviceManagers_.find(node->deviceID);
+
+  if (deviceManagerIt == deviceManagers_.end()) {
+    // Mark the node as no longer executing.
```

Comment:
Isn't this always a bug in our code? Could we log or maybe assert this here?

Reply:
I'll put in an assert for now.